### ISSUE TITLE
RocksDB deleted, but locks still exist

### ISSUE THREAD

**Issue by EliFinkelshteyn** [facebook/rocksdb#2547]
Currently, if I have a connection open to a RocksDB, even if I delete the entire RocksDB on disk, I still can't create a new RocksDB in the same location due to the `No locks available` error. Is this intended behavior, and if so, is there any way to force close all connections to a RocksDB or any other way to force the release of a lock from some connection that wasn't picked up during garbage collection?

**Comment by adamretter** [comment]
Are you using the Java API?

**Comment by EliFinkelshteyn** [comment]
[Python API ](https://github.com/twmht/python-rocksdb), but as far as I can tell, [this part of the code is coming from Rocks DB itself](https://github.com/twmht/python-rocksdb/blob/master/rocksdb/_rocksdb.pyx#L1437).

**Comment by adamretter** [comment]
@EliFinkelshteyn Is this still an issue for you? Which platform are you on please?

**Comment by EliFinkelshteyn** [comment]
@adamretter yes. Seems to be an issue on both OS X and Ubuntu. To get around it, I had to create and have to call a hacky `.close()` function: 

```python
    def close(self):
        if hasattr(self, '_db'):
            del self._db
            gc.collect()
```

**Comment by EliFinkelshteyn** [comment]
Hi @siying,
Thanks for your response. I'm well aware that python-rocksdb is a third party port. I'm not asking questions related to it. I used it merely to demonstrate the issue. I linked you to the portion of the python-rocksdb code that calls out to RocksDB and, in the case of this issue, receives back an error with the statement `No locks available`. 

This `No locks available` error is coming directly from RocksDB and just being relayed by python-rocksdb. It's been about a year since I created the issue and it seems like this language is no longer used in the current version of RocksDB, but you can still see references to it in older issues like [here](https://github.com/facebook/rocksdb/issues/189) (note: this is on RocksDB, not python-rocksdb).

**Comment by gmossessian** [comment]
Hi @siying , thanks for your responses, I work with @EliFinkelshteyn so please excuse my butting in.

It seems that you are correct, that intended Cython behavior is to trigger the C++ destructor upon garbage collection, and there is no access to the `delete` C++ keyword from within Cython. However, I noticed there is a C API for Rocksdb in `c.h`, which includes `extern ROCKSDB_LIBRARY_API void rocksdb_close(rocksdb_t* db);`, would it be appropriate to merely use this method from Cython to close it? 

At this point it's clearly not a rocksdb issue and I am just asking for your advice :) Thank you!

**Comment by siying** [comment]
@gmossessian yes, rocksdb_close() destroy the C++ DB object, so calling it should be sufficient.

**Comment by siying** [comment]
@gmossessian to clarify, if you create DB using C API, then use C API rocksdb_close() to destroy it. If you create it through C++ API, then you have to call C++ `delete` to destroy it. Are you able to do it in CPython?

**Comment by gmossessian** [comment]
wouldn't this would still imply access to C++ from CPython? I was hoping to avoid modifying RocksDB itself, but one way to address this issue that would make everything else easier is to add a `Close()` method to the C++ library, which does not seem to exist as far as I can find it.
